solr/core/src/test/org/apache/solr/cloud/TestClusterProperties.java

```diff
@@ -52,4 +56,88 @@ public void testSetInvalidPluginClusterProperty() throws Exception {
     CollectionAdminRequest.setClusterProperty(propertyName, "valueA")
         .process(cluster.getSolrClient());
   }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testWatchedNodeProperties() {
+    WatchedClusterProperties wcp = new WatchedClusterProperties("node1");
+    NamedList<String> listener1 = new NamedList<>();
+    NamedList<String> all = new NamedList<>();
+    NamedList<String> listener3 = new NamedList<>();
+    wcp.watchNodeProperty("p1", (key, value) -> listener1.add(key, value));
+    wcp.watchNodeProperty(null, (key, value) -> all.add(key, value));
+    BiConsumer<String, String> p3Listener = (key, value) -> listener3.add(key, value);
+    wcp.watchNodeProperty("p3", p3Listener);
+    wcp.onChange(
+        (Map<String, Object>)
+            Utils.fromJSONString(
+                "{\n"
+                    + "  \"watched-node-properties\" :{\n"
```

Comment:
What all way we can define this?

1. Can we differentiate between two components? (say, component-x, component-y)
2. Can we define prop for component-x (node-level, otherwise-all, just-for-all) ?

Reply:
It's not per component, it's just a property per node. May be we can prefix a component name in the property key

Reply:
I think we need three cases here

1.  One can define properties at component level. While registering watch take `component-name`, if same exists then through the exception.
2. Component will register default value, which we can update through ZK (or maybe later nice solr api)
2.1. We can update component1-prop in ZK with node name, saying apply this value for that node. Then we should call listener for that component on that node only.

Side node, recently we introduced this behavior in prod, where we updated max-basic-queries limit on one solr-node. We watched that node for ~4/5 weeks and now we ready to update this limit on all the nodes. This solution require whole cluster restart, that's why we are looking for this feature in solr. Hope this will help.

Reply:
for example , if there is a property ```component-a.property-x``` that component may watch that property explicitly at load time. if no other component has registered a watch on that property, then no callbacks are given to other components. Am I missing something?

Reply:
>>f there is a property component-a.property-x that component may watch that property explicitly at load time.

As long as component is registered, it should get callback. do you think it should register property as well?

Reply:
there is no concept of a "component" is Solr. So, the system would not know whom to give a callback if it does not explicitly register for a callback

Reply:
It's just `name` if we need to use feature flag.  `watch("test1", listener)`